Inserter Menu rendering refactor

notnownikki opened the pull request:
Split the rendering into methods that render a block,
a list of blocks, a category, and all the visible categories.

Removed repeated html and redundant 'addRecent' method.

aduth commented:
Instead of splitting into separate functions, would it make sense to split to their own separate components? Usually distinction is on whether render behavior is conceptually independent / self-contained. "render"-prefixed methods can be a good hint that a component is doing too much work.

Also, I started to look at adding a "No blocks found" search failure message. Would that be made easier or more difficult by this refactoring? Something to keep in mind at least.

notnownikki commented:
I started off making components, I found that passing the callbacks around and dealing with getting references to each block component get a bit messy, it was easier to follow with render functions. To render with a Category and Block component I had to pass callbacks to Category, then pass them on to Block, and to get the refs I had to pass a function to Category, then to Block, which handled the ref, which we need to tab through them. Unless you have another idea on that?

"No blocks found" for search would be about the same amount of work as before I think, we still need to check if there are any visible blocks and render the message if there aren't any, and in the code here and the code before there's only one place we'd need to do that check and render (in this case - renderCategories)

aduth commented on editor/inserter/menu.js:
```diff
@@ -267,14 +264,48 @@ class InserterMenu extends Component {
+				className="editor-inserter__category-blocks"
+				role="menu"
+				tabIndex="0"
+				aria-labelledby={ `editor-inserter__separator-${ separatorSlug }-${ instanceId }` }
```
One issue here, which existed previously as well, is that for the "Recent" tab, we assign an attribute claiming that the block group is labelled by the separator, but there is no separator corresponding to the specified ID. You can see this in the Developer Tools Inspector by searching "editor-inserter__separator-recent-". You'll find only the `aria-labelledby` attribute, but no element with that ID exists.

notnownikki replied:
Left this undefined on the recent tab, where the separators don't render.

aduth replied:
Looking again, it's an issue with "embed" as well. A couple considerations:

- For both recent and embed, we pass `null` as first argument, and implementation of `renderBlocks` detects this in determining whether to assign `aria-labelledby`
- Optionally with above, treat first argument of `renderBlocks` as the ID of labelledby, only specified from `renderCategory`. Maybe this should be the second argument, so in render we can just call `this.renderBlocks( this.props.recentlyUsedBlocks )`

notnownikki replied:
Totally, much more intuitive that way. Fixed.

aduth commented on editor/inserter/menu.js:
```diff
@@ -267,14 +264,48 @@ class InserterMenu extends Component {
+		return (
+			<div
+				className="editor-inserter__category-blocks"
+				role="menu"
```
Do we need `role="menu"` if the parent element includes this already?

notnownikki replied:
fixed!

aduth commented on editor/inserter/menu.js:
```diff
@@ -249,7 +242,11 @@ class InserterMenu extends Component {
 
-	getBlockItem( block ) {
+	switchTab( tab ) {
+		this.setState( { tab: tab } );
```
Minor: ES2015 includes an object shorthand convenience when key and value are the same:

```js
this.setState( { tab } );
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#New_notations_in_ECMAScript_2015

notnownikki replied:
fixed

notnownikki committed:
Inserter Menu rendering refactor

This is to simplify the rendering logic and make it
easier to render arbritary lists of blocks, which
is what we're doing with the recently used blocks.

Split the rendering into methods that render a block,
a list of blocks, a category, and all the visible categories.

Removed repeated html and redundant 'addRecent' method.

Fixed issue with aria-labelledby showing when there was no
element labelling.

notnownikki commented:
Rebased on current master

notnownikki committed:
More sensible arguments for renderBlocks

aduth commented on editor/inserter/menu.js:
```diff
@@ -307,16 +300,54 @@ export class InserterMenu extends Component {
+	}
+
+	renderCategories( visibleBlocksByCategory ) {
+		return getCategories().map(
```
One final thought: Is there not a bit of redundancy in calling `getCategories` when `visibleBlocksByCategory` already includes this information? In other words, could we write this function as:

```js
return mapValues( visibleBlocksByCategory, ( blocks, category ) => (
	this.renderCategory( category, blocks ) )
) );
```

https://lodash.com/docs/4.17.4#mapValues

notnownikki replied:
I don't see how this works... visibleBlocksByCategory doesn't contain all the category information. The data returned by getCategories looks like this:

`{slug: "common", title: "Common Blocks"}`

and visibleBlocksByCategory is just an object with blocks arranged by category slug. We still need the category objects to get the title.

Also, we need to return array of react elements, and mapValues returns an object, so if we did that we'd need to iterate over the keys to render.

aduth replied:
You're right on all accounts 👍  (To the final point, Lodash's `map` supports objects, returning an array, so that could have been an option)

aduth reviewed (approved):
Apologies for letting this one slip the last couple days. I had meant to revisit it sooner 👍
